File: leetcode2020/0020.有效的括号/0020-有效的括号.cpp
class Solution {
public:
    bool isValid(string s) {
        if(s.empty()) return true;
        stack<char> st;
        st.push(s[0]);
        unordered_map<char,char> mp={{')','('},{']','['},{'}','{'}};
        unordered_map<char,char>::iterator it;
        for(int i=1;i<s.size();i++){
            it=mp.find(s[i]);
            if(it!=mp.end()&&!st.empty()&&it->second==st.top()){
                st.pop();
            }else{
                st.push(s[i]);
            }
        }
        if(st.empty()){
            return true;
        }else{
            return false;
        }
    }
};

File: leetcode2020/0020.有效的括号/0020-有效的括号.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        para=[None]
        par={')':'(',']':'[','}':'{'}
        for x in s:
            if x in par and par[x]==para[len(para)-1]:
               para.pop()
            else:
               para.append(x)
        return len(para)==1